notebooks/visit_data.py

```diff
@@ -0,0 +1,255 @@
+fields_of_interest = ['fullVisitorId', 'visitId', 'hour', 'minute', 'pagePath', 'pagePath_clean']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# remove leading '/' if it exists
+visit_data_reduced['pagePath_clean'] = [p[1:] if p[0] == '/' else p for p in visit_data_reduced['pagePath_clean']]
+
+# keep only the first part of the remaining page path
+visit_data_reduced['pagePath_parent'] = [re.split(r'/| |\?|#', p)[0] for p in visit_data_reduced['pagePath_clean']]
+fields_of_interest = ['fullVisitorId', 'visitId', 'hour', 'minute', 'pagePath', 'pagePath_parent']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# find the child page
+fields_of_interest = ['fullVisitorId', 'visitId', 'pagePath_parent', 'pagePath_child']
+visit_data_reduced['pagePath_child'] = [re.split(r'/| |\?|#', p)[-1] for p in visit_data_reduced['pagePath_clean']]
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# concat the parent and child
+pagePath_abbreviated = []
+for parent, child in zip(visit_data_reduced['pagePath_parent'], visit_data_reduced['pagePath_child']):
+    if parent != child:
+        pagePath_abbreviated.append(parent + '/' + child)
+    else:
+        pagePath_abbreviated.append(parent)
+
+visit_data_reduced['pagePath_abbreviated'] = pagePath_abbreviated
+fields_of_interest = ['fullVisitorId', 'visitId', 'pagePath_abbreviated']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# search mask, removes about 4%
+search_mask = visit_data_reduced['pagePath_parent'].str.contains('search')
+visit_data_reduced = visit_data_reduced[~search_mask]
+# empty mask, removes 3.4%
+empty_mask = visit_data_reduced['pagePath_parent'] == ''
+visit_data_reduced = visit_data_reduced[~empty_mask]
+# sign in mask, removes 3.8%
+sign_in_mask = visit_data_reduced['pagePath'].str.contains('sign-in')
+visit_data_reduced = visit_data_reduced[~sign_in_mask]
+# login mask, removes 0.2%
+login_mask = visit_data_reduced['pagePath'].str.contains('login')
+visit_data_reduced = visit_data_reduced[~login_mask]
+# log-in mask, removes 3.2%
+log_in_mask = visit_data_reduced['pagePath'].str.contains('log-in')
+visit_data_reduced = visit_data_reduced[~log_in_mask]
+# done mask, removes 1.0%
+done_mask = visit_data_reduced['pagePath_parent'] == 'done'
+visit_data_reduced = visit_data_reduced[~done_mask]
+
+fields_of_interest = ['fullVisitorId', 'visitId', 'pagePath', 'pagePath_abbreviated']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# lag1 page mask, removes 12%
+visit_data_reduced['pagePath_abbreviated_lag1'] = visit_data_reduced['pagePath_abbreviated'].shift()
+page_mask = visit_data_reduced['pagePath_abbreviated'] != visit_data_reduced['pagePath_abbreviated_lag1']
+visit_data_reduced = visit_data_reduced[page_mask]
```

Comment:
Are you removing those `fullVisitorId` whose `pagePath_abbreviated_lag1` is not equal to `pagePath_abbreviated_lag1`? This means you're losing the first record of users who start their journeys.